spccore/internal/cache.py

```diff
@@ -0,0 +1,328 @@
+        validate_type(str, file_name, "file_name")
+
+        cache_dir = self._get_cache_dir(file_handle_id)
+        if not os.path.exists(cache_dir):
+            return None
+        if file_path is None:
+            file_path = self.get_default_file_path(file_handle_id, file_name)
+        if os.path.exists(file_path) and not _is_modified(cache_dir, file_path):
+            return os.path.abspath(file_path)
+        return None
+
+    def get_all_cached_file_path(self, file_handle_id: int) -> list:
+        """
+        Retrieve all file paths for the given file handle id from the cache
+
+        :param file_handle_id: the file handle id to look up
+        :returns: The paths to the file in the cache if the file exists and has not been modified
+        """
+        validate_type(int, file_handle_id, "file_handle_id")
+
+        cache_dir = self._get_cache_dir(file_handle_id)
+        if not os.path.exists(cache_dir):
+            return list()
+        return _get_all_non_modified_paths(cache_dir)
+
+    def register(self, file_handle_id: int, file_path: str) -> dict:
+        """
+        Register file_path with file_handle_id in the cache
+
+        :param file_handle_id: the file handle id of the file
+        :param file_path: the actual file path
+        :return: the cache map
+        """
+        validate_type(int, file_handle_id, "file_handle_id")
+        validate_type(str, file_path, "file_path")
+
+        if not file_path or not os.path.exists(file_path):
+            raise ValueError("Can't find file \"%s\"" % file_path)
+
+        cache_dir = self._get_cache_dir(file_handle_id)
+        with Lock(SYNAPSE_DEFAULT_CACHE_MAP_FILE_NAME, current_working_directory=cache_dir):
+            cache_map = _get_cache_map(cache_dir)
+            file_path = _normalize_path(file_path)
+            cache_map[file_path] = _get_modified_time_in_iso(file_path)
+            _write_cache_map(cache_map, cache_dir)
+
+        return cache_map
+
+    def remove(self, file_handle_id: int, *, file_path: str = None, delete_file: bool = False) -> list:
+        """
+        Remove a file(s) from the cache.
+
+        :param file_handle_id: the file handle id to look up
+        :param file_path: the file_path to look up and remove.
+            Default None, which will remove all copies found in the cache.
+        :param delete_file: Set to True to remove the actual file. Default False.
+
+        :returns: A list of files removed from the cache.
+        """
+        validate_type(int, file_handle_id, "file_handle_id")
+        removed = []
+        cache_dir = self._get_cache_dir(file_handle_id)
+
+        with Lock(SYNAPSE_DEFAULT_CACHE_MAP_FILE_NAME, current_working_directory=cache_dir):
+            cache_map = _get_cache_map(cache_dir)
+
+            if file_path is None:
+                for path in cache_map:
+                    if delete_file is True and os.path.exists(path):
+                        os.remove(path)
```

Comment:
avoid doing expensive IO operations like deleting files while holding the lock.  It may be fine to delete the file prior to obtaining the lock